tribits/core/package_arch/TribitsProjectWriteConfigFile.cmake

```diff
@@ -0,0 +1,228 @@
+  set(PACKAGE_LIST ${${PROJECT_NAME}_DEFINED_INTERNAL_PACKAGES})
+  if (PACKAGE_LIST)
+    list(REVERSE PACKAGE_LIST)
+  endif()
+
+  # Loop over all packages to determine which were enabled. Then build a list
+  # of all their libraries/includes in the proper order for linking
+  set(FULL_PACKAGE_SET "")
+  set(FULL_LIBRARY_SET "")
+  foreach(TRIBITS_PACKAGE ${PACKAGE_LIST})
+    if(${PROJECT_NAME}_ENABLE_${TRIBITS_PACKAGE})
+      list(APPEND FULL_PACKAGE_SET ${TRIBITS_PACKAGE})
+      list(APPEND FULL_LIBRARY_SET ${${TRIBITS_PACKAGE}_LIBRARIES})
+    endif()
+  endforeach()
+
+  set(${PROJECT_NAME}_CONFIG_LIBRARIES ${FULL_LIBRARY_SET})
+
+  # Reversing the tpl list so that the list of tpls will be produced in
+  # order of most dependent to least dependent.
+  if (${PROJECT_NAME}_DEFINED_TPLS)
+    set(TPL_LIST ${${PROJECT_NAME}_DEFINED_TPLS})
+    list(REVERSE TPL_LIST)
+  endif()
+
+  # Loop over all TPLs to determine which were enabled. Then build a list
+  # of all their libraries/includes in the proper order for linking
+  set(FULL_TPL_SET "")
+  set(FULL_TPL_LIBRARY_SET "")
+  foreach(TPL ${TPL_LIST})
+    if(TPL_ENABLE_${TPL})
+      list(APPEND FULL_TPL_SET ${TPL})
+      list(APPEND FULL_TPL_LIBRARY_SET ${TPL_${TPL}_LIBRARIES})
+    endif()
+  endforeach()
+
+  set(${PROJECT_NAME}_CONFIG_TPL_LIBRARIES ${FULL_TPL_LIBRARY_SET})
+
+  #
+  # Configure two files for finding ${PROJECT_NAME}. One for the build tree
+  # and one for installing
+  #
+
+  # Generate a note discouraging editing of the <package>Config.cmake file
+  set(DISCOURAGE_EDITING "Do not edit: This file was generated automatically by CMake.")
+
+  # Write the specification of the rpath if necessary. This is only needed if
+  # we're building shared libraries.
+  if(BUILD_SHARED_LIBS)
+    string(REPLACE ";" ":" SHARED_LIB_RPATH_COMMAND
+     "${${PROJECT_NAME}_CONFIG_LIBRARY_DIRS}")
+    set(SHARED_LIB_RPATH_COMMAND ${CMAKE_SHARED_LIBRARY_RUNTIME_CXX_FLAG}${SHARED_LIB_RPATH_COMMAND})
+  endif()
+
+  # Custom code in configuration file.
+  set(PROJECT_CONFIG_CODE "")
+
+  # Appending the logic to include each package's config file.
+  set(LOAD_CODE "# Load configurations from enabled packages")
+  foreach(TRIBITS_PACKAGE ${FULL_PACKAGE_SET})
+    set(LOAD_CODE "${LOAD_CODE}
+include(\"${${TRIBITS_PACKAGE}_BINARY_DIR}/${TRIBITS_PACKAGE}Config.cmake\")")
+  endforeach()
+  set(PROJECT_CONFIG_CODE "${PROJECT_CONFIG_CODE}\n${LOAD_CODE}")
+
+  tribits_set_compiler_vars_for_config_file(INSTALL_DIR)
+
+  if (${PROJECT_NAME}_ENABLE_INSTALL_CMAKE_CONFIG_FILES)
+    # In TribitsProjectConfigTemplate.cmake.in, we would like to preserve
+    # ${}-variables after the conversion to TribitsProjectConfigTemplate.cmake.
+    # To this end, one typically uses the @-syntax for variables. That doesn't
+    # support nested variables, however.  Use ${PDOLLAR} as a workaround, cf.
+    # <http://www.cmake.org/pipermail/cmake/2013-April/054341.html>.
+    set(PDOLLAR "$")
```

Comment:
Here is where you see `PDOLLAR` gets used:

```
$ cd TriBITS
$ find tribits -type f -exec grep -l "[^a-zA-Z0-9_]PDOLLAR[^a-zA-Z0-9_]" {} \;
tribits/core/package_arch/TribitsProjectWriteConfigFile.cmake
tribits/core/package_arch/TribitsInternalPackageWriteConfigFile.cmake
tribits/core/installation/TribitsProjectConfigTemplate.cmake.in
tribits/core/installation/TribitsPackageConfigTemplate.cmake.in
```

I did not write this code.  It seems to have come from:

* https://cmake.org/pipermail/cmake/2013-April/054353.html

Using `configure_file(... @ONLY)` is likely a better solution but there may be some gotcha that I don't see here.  The author of that code must have considered that because it was discussed in the CMake email thread at:

* https://cmake.org/pipermail/cmake/2013-April/054341.html

This code and the generation of these files could likely use some more cleanup.  (I did a lot of simplifications of the generation code in earlier PRs.)